pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,452 @@
+
+		t.connsLock.Lock()
+		defer t.connsLock.Unlock()
+
+		t.logger.Log(logging.LevelDebug, "started connecting nodes", "src", t.ownID)
+		defer t.logger.Log(logging.LevelDebug, "finished connecting nodes", "src", t.ownID)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(len(nodes))
+
+		for nodeID := range nodes {
+			// Do not establish a real connection with own node.
+			if nodeID == t.ownID {
+				wg.Done()
+				continue
+			}
+
+			go t.connectToNodeWithoutLock(ctx, nodeID, wg)
+		}
+		wg.Wait()
+	}()
+}
+
+func (t *Transport) reconnect(ctx context.Context, node types.NodeID) {
+	info, found := t.getNodeAddrInfo(node)
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", node)
+	} else {
+		if err := t.host.Network().ClosePeer(info.ID); err != nil {
+			t.logger.Log(logging.LevelError, "could not close the connection to node", "src", t.ownID, "dst", node, "err", err)
+		}
+
+		t.connWg.Add(1)
+		go func() {
+			defer t.connWg.Done()
+
+			t.connsLock.Lock()
+			defer t.connsLock.Unlock()
+
+			t.connWg.Add(1)
+			t.connectToNodeWithoutLock(ctx, node, t.connWg)
+		}()
+	}
+}
+
+func (t *Transport) connectToNodeWithoutLock(ctx context.Context, nodeID types.NodeID, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	conn, found := t.conns[nodeID]
+	if found &&
+		t.host.Network().Connectedness(t.conns[nodeID].AddrInfo.ID) == network.Connected {
+		t.logger.Log(logging.LevelInfo, "connection to node already exists", "src", t.ownID, "dst", nodeID)
+		return
+	}
+
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", nodeID)
+		return
+	}
+
+	info := conn.AddrInfo
+
+	t.host.Peerstore().AddAddrs(info.ID, info.Addrs, PermanentAddrTTL)
+
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v is connecting to node %v", t.ownID, nodeID))
+	s, err := t.openStream(ctx, info.ID)
+	if err != nil {
+		t.logger.Log(logging.LevelError, "failed to open stream to node", "addr", info, "node", nodeID, "err", err)
+		return
+	}
+
+	conn.Stream = s
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), nodeID.Pb()))
+}
+
+func (t *Transport) openStream(ctx context.Context, dest peer.ID) (network.Stream, error) {
+	// We need the simplest retry mechanism due to the fact that the underlying libp2p's NewStream function dials once:
+	// https://github.com/libp2p/go-libp2p/blob/7828f3e0797e0a7b7033fa5e8be9b94f57a4c173/p2p/net/swarm/swarm.go#L358
+	t.logger.Log(logging.LevelDebug, "opening stream to peer", "src", t.ownID, "dst", dest)
+
+	// The implementation is based on the openStream() function from the RemoteTracer:
+	// https://github.com/libp2p/go-libp2p-pubsub/blob/cbb7bfc1f182e0b765d2856f6a0ea73e34d93602/tracer.go#L280
+	var s network.Stream
+	var err error
+	for i := 0; i < maxRetries; i++ {
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, err = t.host.NewStream(sctx, dest, ProtocolID)
+		cancel()
```

Comment:
Why do you call `cancel` here?

Reply:
Because that is context.WithTimeout's cancel. 
```
Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this
```